tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -561,26 +561,45 @@ EnqueueRecordEventCommand::EnqueueRecordEventCommand(
+
+    uint8_t num_hw_cqs = this->device->num_hw_cqs(); // Device initialize asserts that there can only be a maximum of 2 HW CQs
+    uint32_t packed_event_payload_sizeB = align(sizeof(CQDispatchCmd) + num_hw_cqs * sizeof(CQDispatchWritePackedUnicastSubCmd), L1_ALIGNMENT)
+        + (align(EVENT_PADDED_SIZE, L1_ALIGNMENT) * num_hw_cqs);
+    uint32_t packed_write_sizeB = align(sizeof(CQPrefetchCmd) + packed_event_payload_sizeB, PCIE_ALIGNMENT);
+
     uint32_t cmd_sequence_sizeB = CQ_PREFETCH_CMD_BARE_MIN_SIZE + // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WAIT
-        align(CQ_PREFETCH_CMD_BARE_MIN_SIZE + EVENT_PADDED_SIZE, PCIE_ALIGNMENT) + // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WRITE_LINEAR + event ID
+        packed_write_sizeB + // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WRITE_PACKED + unicast subcmds + event payload
         align(CQ_PREFETCH_CMD_BARE_MIN_SIZE + EVENT_PADDED_SIZE, PCIE_ALIGNMENT); // CQ_PREFETCH_CMD_RELAY_INLINE + CQ_DISPATCH_CMD_WRITE_LINEAR_HOST + event ID
 
     DeviceCommand command_sequence(cmd_sequence_sizeB);
 
     command_sequence.add_dispatch_wait(false, DISPATCH_MESSAGE_ADDR, this->expected_num_workers_completed);
 
-    std::vector<uint32_t> event_payload(EVENT_PADDED_SIZE / sizeof(uint32_t), 0);
-    event_payload[0] = this->event_id;
-
-    uint8_t num_hw_cqs = this->device->num_hw_cqs();
-    uint16_t channel = tt::Cluster::instance().get_assigned_channel_for_device(this->device->id());
-    tt_cxy_pair dispatch_location = dispatch_core_manager::get(num_hw_cqs).dispatcher_core(this->device->id(), channel, this->command_queue_id);
     CoreType core_type = dispatch_core_manager::get(num_hw_cqs).get_dispatch_core_type(this->device->id());
-    CoreCoord dispatch_physical_core = get_physical_core_coordinate(dispatch_location, core_type);
+    uint16_t channel = tt::Cluster::instance().get_assigned_channel_for_device(this->device->id());
+    std::vector<CQDispatchWritePackedUnicastSubCmd> unicast_sub_cmds(num_hw_cqs);
+    std::vector<const void *> event_payloads(num_hw_cqs);
+
+    for (uint8_t cq_id = 0; cq_id < num_hw_cqs; cq_id++) {
+        tt_cxy_pair dispatch_location = dispatch_core_manager::get(num_hw_cqs).dispatcher_core(this->device->id(), channel, cq_id);
+        CoreCoord dispatch_physical_core = get_physical_core_coordinate(dispatch_location, core_type);
+        unicast_sub_cmds[cq_id] = CQDispatchWritePackedUnicastSubCmd{.noc_xy_addr = get_noc_unicast_encoding(dispatch_physical_core)};
+        event_payloads[cq_id] = event_payload.data();
+    }
 
-    bool flush_prefetch = true;
-    command_sequence.add_dispatch_write_linear(flush_prefetch, 0, get_noc_unicast_encoding(dispatch_physical_core), CQ_COMPLETION_LAST_EVENT, EVENT_PADDED_SIZE, event_payload.data());
+    uint32_t address = this->command_queue_id == 0 ? CQ0_COMPLETION_LAST_EVENT : CQ1_COMPLETION_LAST_EVENT;
+    command_sequence.add_dispatch_write_packed<CQDispatchWritePackedUnicastSubCmd>(
```

Comment:
Write packed writes event ID to all dispatch cores that service active command queues on a single device. 

For single CQ, dispatch will write to its local L1. For 2 CQs dispatch core servicing the EnqueueRecordEvent will also write to dispatch core servicing the other CQ. 

Each dispatch core has one location for last event in CQ0 and another for last event in CQ1. This allows EnqueueWaitForEvent to poll its local L1